### PR TITLE
Make BenchmarkModule compatible with PyTorch Lightning 2.0

### PR DESCRIPTION
**Pull request by guarin**
## What has changed and why?

* Make BenchmarkModule compatible with PyTorch Lightning 2.0
* Create KNN features at start of validation epoch instead of end of training epoch

The benchmarks are not yet completely compatible with PyTorch Lightning 2.0 as we use the LARS optimizer from PyTorch Lightning Bolts which is not compatible with PyTorch Lightning 2.0: https://github.com/Lightning-Universe/lightning-bolts/issues/962

I guess the simplest solution for this is to copy one of the LARS implementations into our package as the code is pretty simple and we can avoid additional dependencies.

## How was it tested?
* Ran imagenette benchmark on simclr and got same accuracy as before
* Tested on PyTorch Lightning 1.8.0 and 2.0.1

### CODE REVIEW

File: lightly/utils/benchmarking.py [hunk 1]
```diff
@@ -171,59 +174,61 @@ def __init__(
...
-                self.targets_bank,
+                self._train_features,
+                self._train_targets,
                 self.num_classes,
                 self.knn_k,
                 self.knn_t,
             )
-            num = images.size()
-            top1 = (pred_labels[:, 0] == targets).float().sum()
-            return (num, top1)
```
[the remark this answers is not in the document]
Comment (MalteEbner) [reply lightly/utils/benchmarking.py:211]: Why is it moved instead into the `on_validation_epoch_end`?

**Review by MalteEbner — COMMENTED** [review]
I don't understand the design of this fully yet and why which code is distributed across
`on_validation_epoch_start`
`validation_step`
`on_validation_epoch_end`

I am also a bit unsure about the usage.
E.g. before the change, the top1 was returned by the. `validation_step`.
Now this `self.max_accuracy` is set by `on_validation_epoch_end`.
Is this required by the new pytorch 2.0?

Furthermore, I am wondering whether we have a unittest for parts of it.

File: lightly/utils/benchmarking.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/utils/benchmarking.py:211]: > Why is it moved instead into the on_validation_epoch_end?

Because PyTorch Lightning 2.0 removed the `training_epoch_end ` and `validation_epoch_end ` methods.

> Why is nothing returned anymore?

Previously, PyTorch Lightning would aggregate all returned values from `validation_step` and pass them to `validation_epoch_end` as the `outputs` argument. This was removed in 2.0 and now it is up to users to keep track of the values.

**Comment by guarin** [comment]
> I don't understand the design of this fully yet and why which code is distributed across
on_validation_epoch_start
validation_step
on_validation_epoch_end

* `on_validation_epoch_start` -> Generate KNN features
* `validation_step` -> Get nearest neighbors for current validation batch
* `on_validation_epoch_end` -> Calculate accuracy

We use `on_validation_epoch_start` instead of `training_epoch_end` or `training_epoch_end` because we only want to calculate the knn features when we actually do validation.

File: tests/utils/test_benchmarking.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        model = _DummyModel(dataloader_kNN=dataloader_train, knn_k=3)
+        trainer = Trainer(max_epochs=2)
+        trainer.fit(
+            model,
+            train_dataloaders=dataloader_train,
+            val_dataloaders=dataloader_val,
+        )
+        assert (
+            model.max_accuracy < 1.0
+        )  # accuracy is <1.0 because train val are different
```
Comment (MalteEbner) [inline tests/utils/test_benchmarking.py:66]: Without seeding, I would not rely on that.

**Review by MalteEbner — APPROVED** [review]
Makes sense to me, thank you for explaining it.

File: tests/utils/test_benchmarking.py (hunk 2 above)
Comment (guarin) [reply tests/utils/test_benchmarking.py:66]: There is a seed, we set `torch.manual_seed(0)` at the top of the method.
